torchtext/modules/multiheadattention.py

```diff
@@ -0,0 +1,213 @@
+
+
+class ScaledDotProduct(torch.nn.Module):
+
+    def __init__(self, dropout=0.0):
+        r"""Processes a projected query and key-value pair to apply
+        scaled dot product attention.
+
+        Args:
+            dropout (float): probability of dropping an attention weight.
+
+        Examples::
+            >>> SDP = torchtext.models.ScaledDotProduct(0.1)
+            >>> q = torch.randn(256, 21, 3)
+            >>> k = v = torch.randn(256, 21, 3)
+            >>> attn_output, attn_weights = SDP(q, k, v)
+            >>> print(attn_output.shape, attn_weights.shape)
+            torch.Size([256, 21, 3]) torch.Size([256, 21, 21])
+        """
+        super(ScaledDotProduct, self).__init__()
+        self.dropout = dropout
+
+    def forward(self, query: torch.Tensor, key: torch.Tensor, value: torch.Tensor,
+                attn_mask: Optional[torch.Tensor] = None,
+                bias_k: Optional[torch.Tensor] = None,
+                bias_v: Optional[torch.Tensor] = None) -> Tuple[torch.Tensor, torch.Tensor]:
+        r"""Uses a scaled dot product with the projected key-value pair to update
+        the projected query.
+
+        Args:
+            query (Tensor): Projected query
+            key (Tensor): Projected key
+            value (Tensor): Projected value
+            attn_mask (BoolTensor, optional): 3D mask that prevents attention to certain positions.
+            bias_k and bias_v: (Tensor, optional): one more key and value sequence to be added at
+                sequence dim (dim=-3). Those are used for incremental decoding. Users should provide
+                non-None to both arguments in order to activate them.
+
+        Shape:
+            - query: :math:`(L, N * H, E / H)`
+            - key: :math:`(S, N * H, E / H)`
+            - value: :math:`(S, N * H, E / H)`
+            - attn_mask: :math:`(N * H, L, S)`, positions with ``True`` are not allowed to attend
+                while ``False`` values will be unchanged.
+            - bias_k and bias_v:bias: :math:`(1, N * H, E / H)`
+
+            - Output: :math:`(L, N * H, E / H)`, :math:`(N * H, L, S)`
+
+            where L is the target length, S is the source length, H is the number
+            of attention heads, N is the batch size, and E is the embedding dimension.
+        """
+        if bias_k is not None and bias_v is not None:
+            assert key.size(-1) == bias_k.size(-1) and key.size(-2) == bias_k.size(-2) and bias_k.size(-3) == 1, \
+                "Shape of bias_k is not supported"
+            assert value.size(-1) == bias_v.size(-1) and value.size(-2) == bias_v.size(-2) and bias_v.size(-3) == 1, \
+                "Shape of bias_v is not supported"
+            key = torch.cat([key, bias_k])
+            value = torch.cat([value, bias_v])
+            if attn_mask is not None:
+                _attn_mask = attn_mask
+                attn_mask = torch.nn.functional.pad(_attn_mask, (0, 1))
```

Comment:
@zhangguanheng66 Why not simply `attn_mask = torch.nn.functional.pad(attn_mask, (0, 1))`?

Reply:
Update in the revised_mha PR (link])(https://github.com/pytorch/text/pull/839/commits/10fb9583c0bc46b229323766555f9b089e51441d)